### PR TITLE
To Fix socket close on QUIC pipe

### PR DESCRIPTION
**Pull request by JaylinYu**
make fini work of QUIC transport working

### CODE REVIEW

**Comment by JaylinYu** [comment]
still need test

**Review by lee-emqx — APPROVED** [review]

File: src/supplemental/quic/msquic_dial.c [hunk 1]
```diff
@@ -211,6 +211,11 @@ quic_dialer_cb(void *arg)
 	}
 
 	c = d->currcon;
+	if (c == NULL) {
+		// abort this cb
+		nni_mtx_unlock(&d->mtx);
+		return;
+	}
```
Comment (wanghaEMQ) [inline src/supplemental/quic/msquic_dial.c:218]: It should not happen. Did you ever meet the case that c is NULL?

File: src/supplemental/quic/msquic_dial.c [hunk 2]
```diff
@@ -490,24 +498,27 @@ quic_error(void *arg, int err)
 
 	nni_mtx_lock(&c->mtx);
 	// only close aio of this stream
-	while (((aio = nni_list_first(&c->readq)) != NULL) ||
-	    ((aio = nni_list_first(&c->writeq)) != NULL)) {
+	while ((aio = nni_list_first(&c->writeq)) != NULL) {
+		nni_aio_list_remove(aio);
+		nni_aio_finish_error(aio, err);
+	}
+	while ((aio = nni_list_first(&c->readq)) != NULL) {
```
Comment (wanghaEMQ) [inline src/supplemental/quic/msquic_dial.c:505]: Emmm I can't get any difference from the code before

File: src/supplemental/quic/msquic_dial.c [hunk 3]
```diff
@@ -517,12 +528,12 @@ quic_close2(void *arg)
...
-		    ((aio = nni_list_first(&c->writeq)) != NULL)) {
-			nni_aio_list_remove(aio);
-			nni_aio_finish_error(aio, NNG_ECLOSED);
-		}
-		msquic_strm_fini(c->qstrm);
+		// while (((aio = nni_list_first(&c->readq)) != NULL) ||
+		//     ((aio = nni_list_first(&c->writeq)) != NULL)) {
+		// 	nni_aio_list_remove(aio);
+		// 	nni_aio_finish_error(aio, NNG_ECLOSED);
+		// }
```
Comment (wanghaEMQ) [inline src/supplemental/quic/msquic_dial.c:535]: Why?

File: src/mqtt/protocol/mqtt/mqtt_quic_client.c [hunk 4]
```diff
@@ -1429,6 +1428,22 @@ mqtt_quic_sock_get_sqlite_option(mqtt_sock_t *s)
...
+	bool              b;
+
+	if (((rv = nni_copyin_bool(&b, buf, sz, t)) != 0)) {
+		return (rv);
+	}
+	nni_mtx_lock(&s->mtx);
+	s->multi_stream = b;
+	nni_mtx_unlock(&s->mtx);
+	return (0);
+}
```
Comment (wanghaEMQ) [inline src/mqtt/protocol/mqtt/mqtt_quic_client.c:1445]: Setting option multi-stream was supported. And relative code is in quic_api.c

**Comment by wanghaEMQ** [comment]
https://github.com/nanomq/NanoNNG/issues/686

File: src/supplemental/quic/msquic_dial.c (hunk 3 above)
Comment (JaylinYu) [reply src/supplemental/quic/msquic_dial.c:535]: make it unique

File: src/supplemental/quic/msquic_dial.c (hunk 2 above)
Comment (JaylinYu) [reply src/supplemental/quic/msquic_dial.c:505]: actually you are right

**Review by JaylinYu — COMMENTED** [review]
only quic

**Pull request merged**